File: packages/examples/todo/backends/rust/src/commands/toggle.rs
use super::{not_found, ok, schema, Command};
use crate::store::TodoStore;
use afd::{failure, CommandResult};
use serde::Deserialize;
use serde_json::{json, Value};
use std::sync::Arc;

pub struct Toggle(pub Arc<TodoStore>);

#[derive(Deserialize)]
pub struct Input {
    id: String,
}

impl Command for Toggle {
    type Input = Input;
    const NAME: &'static str = "todo-toggle";
    const DESCRIPTION: &'static str = "Toggle a todo's completion status";
    const MUTATION: bool = true;

    fn schema() -> Value {
        json!({
            "type": "object",
            "properties": { "id": schema::id() },
            "required": ["id"]
        })
    }

    fn run(&self, input: Input) -> CommandResult<Value> {
        match self.0.toggle(&input.id) {
            Some(todo) => {
                let action = if todo.completed {
                    "Marked as complete"
                } else {
                    "Marked as incomplete"
                };
                let reasoning = format!("{action}: \"{}\"", todo.title);
                ok(&todo, reasoning, 1.0, Vec::new())
            }
            None => failure(not_found(&input.id)),
        }
    }
}

File: packages/experiments/in-process-binding/backend/src/commands/toggle.ts
/**
 * @fileoverview todo-toggle command
 */

import { defineCommand, failure, success } from '@lushly-dev/afd-server';
import { z } from 'zod';
import { store } from '../store/index.js';
import type { Todo } from '../types.js';

const inputSchema = z.object({
	id: z.string().min(1, 'Todo ID is required'),
});

export const toggleTodo = defineCommand<typeof inputSchema, Todo>({
	name: 'todo-toggle',
	description: 'Toggle the completion status of a todo',
	category: 'todo',
	tags: ['todo', 'toggle', 'write', 'single'],
	mutation: true,
	version: '1.0.0',
	input: inputSchema,
	errors: ['NOT_FOUND'],

	async handler(input) {
		const existing = store.get(input.id);
		if (!existing) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${input.id}" not found`,
				suggestion: 'Use todo-list to see available todos',
			});
		}

		const updated = store.toggle(input.id);

		if (!updated) {
			return failure({
				code: 'NOT_FOUND',
				message: `Todo with ID "${input.id}" not found`,
				suggestion: 'Use todo-list to see available todos',
			});
		}

		const action = updated.completed ? 'Marked as completed' : 'Marked as pending';

		return success(updated, {
			reasoning: `${action}: "${updated.title}"`,
			confidence: 1.0,
		});
	},
});
